File: Homework/subscription.py
import datetime
import json


class Subscription:
    def __init__(self, company: list = None, value: list = None, drop: list = None, variation: list = None, date: list = None):
        self.company = company
        self.value = value
        self.drop = drop
        self.variation = variation
        self.date = date

    def to_json(self):
        # create a json object from the object with the attributes that are not None
        # return the json object
        json_dict = {}
        if self.company:
            json_dict['company'] = self.company

        if self.value:
            json_dict['value'] = self.value

        if self.drop:
            json_dict['drop'] = self.drop

        if self.variation:
            json_dict['variation'] = self.variation

        if self.date:
            json_dict['date'] = self.date #.strftime('%d-%m-%Y')

        return json_dict
    
    @staticmethod
    def from_json(json_dict):
        # create a new object from the json object
        # return the new object
        #json_dict = json.loads(json_string)
        return Subscription(
            company=json_dict.get("company"),
            value=json_dict.get("value"),
            drop=json_dict.get("drop"),
            variation=json_dict.get("variation"),
            date=json_dict.get("date")
        )

    def print_fields(self):
        print("company: ",self.company, ";",
              "value: ",self.value, ";",
              "drop: ",self.drop, ";",
              "variation: ",self.variation, ";",
              "date: ",self.date)

File: POC_RabbitMq/src/main/java/ebs/generator/entities/Subscription.java
package ebs.generator.entities;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import ebs.generator.Utils;
import org.jetbrains.annotations.NotNull;
import org.json.JSONObject;

import java.sql.Date;

import static ebs.generator.entities.Constants.*;

@Getter
@NoArgsConstructor
@AllArgsConstructor
public class Subscription {
    private Pair<String, String> company    = null;
    private Pair<String, Float> value       = null;
    private Pair<String, Float> drop        = null;
    private Pair<String, Float> variation   = null;
    private Pair<String, Date> date         = null;

    public void setCompany(String sign) {
        this.company = new Pair<>(sign, Utils.generateRandomCompany());
    }

    public void setValue() {
        this.value = new Pair<>(
                Utils.generateRandomSign(2, 6),
                Utils.generateRandomFloat(MIN_VALUE, MAX_VALUE)
        );
    }

    public void setDrop() {
        this.drop = new Pair<>(
                Utils.generateRandomSign(2, 6),
                Utils.generateRandomFloat(MIN_DROP, MAX_DROP)
        );
    }

    public void setVariation() {
        this.variation = new Pair<>(
                Utils.generateRandomSign(2, 6),
                Utils.generateRandomFloat(MIN_VARIATION, MAX_VARIATION)
        );
    }

    public void setDate() {
        this.date = new Pair<>(
                Utils.generateRandomSign(2, 6),
                Utils.generateRandomDate(NUMBER_OF_DAYS_AGO)
        );
    }

    public boolean isAttributeSet(int atr){
        return switch (atr) {
            case 1 -> company != null;
            case 2 -> value != null;
            case 3 -> drop != null;
            case 4 -> variation != null;
            case 5 -> date != null;
            default -> false;
        };
    }

    public JSONObject toJson(){
        return new JSONObject() {{
            if (company     != null)  put("company", company.toJson());
            if (value       != null)  put("value", value.toJson());
            if (drop        != null)  put("drop", drop.toJson());
            if (variation   != null)  put("variation", variation.toJson());
            if (date        != null)  put("date", date.toJson());
        }};
    }

    public static @NotNull Subscription fromJson(@NotNull JSONObject json){
        Subscription subscription = new Subscription();
        if (json.has("company")) {
            subscription.company = Pair.fromJson(json.getJSONObject("company"));
        }
        if (json.has("value")) {
            subscription.value = Pair.fromJson(json.getJSONObject("value"));
        }
        if (json.has("drop")) {
            subscription.drop = Pair.fromJson(json.getJSONObject("drop"));
        }
        if (json.has("variation")) {
            subscription.variation = Pair.fromJson(json.getJSONObject("variation"));
        }
        if (json.has("date")) {
            subscription.date = Pair.fromJson(json.getJSONObject("date"));
        }
        return subscription;
    }

    @Override
    public String toString() {
        return toJson().toString();
    }

    public boolean matches(JSONObject jsonObject) {

        Subscription other = Subscription.fromJson(jsonObject);

        if (company != null && !company.equals(other.company)) return false;
        if (value != null && !value.equals(other.value)) return false;
        if (drop != null && !drop.equals(other.drop)) return false;
        if (variation != null && !variation.equals(other.variation)) return false;
        if (date != null && !date.equals(other.date)) return false;

        return true;
    }

    @Override
    public int hashCode() {
        return toString().hashCode();
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof Subscription) {
            return this.toString().equals(obj.toString());
        }
        return false;
    }
}
